### ISSUE TITLE
JSON Output

### ISSUE THREAD

**Issue by paulstaab** [scrm/scrm#86]
I'd like to add an option to produce JSON formated output as an alternative to the standard ms style output. This would then be easy to import from almost any programming language. This could look like:

```
{
  "program": "scrm",
  "version": "1.7.0",
  "arguments": "5 1 -r 1 100 -T -L -t 5",
  "seed": 200123512,
  "results": [
    { "segments": [
        { "length": 17.3, "newick_tree": "((1:0...)", "tmrca": 0.81, "tree_length": 1.33 },
        { "length": 82.7, "newick_tree": "((1:0...)", "tmrca": 0.65, "tree_length": 1.08 },
      ], 
      "mutations": [
        { "position": 12.02, "time": 0.65, "sample": [0, 0, 1, 1, 0] },
        { "position": 27.05, "time": 0.45, "sample": [1, 0, 0, 0, 0] },
        ...
      ]
    }
  ]
}
```

@shajoezhu @gerton what is your opinion here? This shouldn't be too hard to implement.

@jeromekelleher Are you interested in defining a common format?

**Comment by shajoezhu** [comment]
I think it is a great idea. please let me know which part you want me to work on

**Comment by jeromekelleher** [comment]
This is a great idea @paulstaab, and I agree we should start moving away from ms's output. However, I worry about how this JSON would scale. For large simulations, this is going to be very, very large (many terabytes for 100k samples over a chromosome). Also, because it's a single JSON document it's going to be hard to parse incrementally, and in practise the user will need to parse the JSON by hand to get the bits they want.

I've been working on an HDF5 based format which is based on the backwards-in-time algorithm, but should also be applicable to left-to-right algorithms. It's early days, and I haven't fully written up how this is derived, but the user documentation is here: https://msprime.readthedocs.org/en/latest/file-format.html. The big advantages are conciseness and efficiency of tree access: for large simulations, we move from several terabytes in newick to roughly 100 megabytes, and processing the trees is very fast. Because of this, I don't see a lot of value in any format based on newick.

**Comment by gerton** [comment]
Hi Paul — I think that’s an excellent idea; JSON support is very good in many languages, and it’s relatively fast to parse as well.  It would make life easier for many power-users I think.

BW
G
